### PR TITLE
More dominators-related utilities

### PR DESCRIPTION
**Pull request by xclerc**
This pull request moves the existing computation of
dominators to a dedicated module, and adds the
following:
- immediate dominators;
- dominance frontier;
- dominator tree.

### CODE REVIEW

**Comment by xclerc** [comment]
> I haven't reviewed the code that moved from [cfg_loop_infos.ml](https://github.com/ocaml-flambda/flambda-backend/pull/1440/files#diff-ae1ceb73bb6b64db0742c21b21fedf6ca54ade05e52241943b3e1ff4a1297333) to `dominators.ml`, assuming it hasn't changed.

Indeed it has not changed.

File: backend/cfg/cfg_dominators.mli [hunk 1]
```diff
@@ -8,22 +8,32 @@ val compute_dominators : Cfg.t -> dominators
...
 
 type immediate_dominators = Label.t Label.Map.t
 
 val compute_immediate_dominators : Cfg.t -> dominators -> immediate_dominators
+(* [compute_immediate_dominators cfg doms] returns a map from labels to their
+   immediate dominator, using the following definition: "the immediate dominator
+   a node n is the unique node that strictly dominates n but does not strictly
+   dominate any other node that strictly dominates n". By definition, the
+   returned map has an binding for each node of [cfg], except the entry
+   point. *)
```
Comment (gretay-js) [inline backend/cfg/cfg_dominators.mli:27]:
```suggestion
(* [compute_immediate_dominators cfg doms] returns a map from labels to their
   immediate dominator, using the following definition: "the immediate dominator
   of a node n is the unique node that strictly dominates n but does not strictly
   dominate any other node that strictly dominates n". By definition, the
   returned map has a binding for each node of [cfg], except the entry
   point, assuming that all nodes are reachable from entry. *)
```

File: backend/cfg/cfg_dominators.mli [hunk 2]
```diff
@@ -8,22 +8,32 @@ val compute_dominators : Cfg.t -> dominators
...
 
 type dominance_frontiers = Label.Set.t Label.Map.t
 
 val compute_dominance_frontiers :
   Cfg.t -> dominators -> immediate_dominators -> dominance_frontiers
+(* [compute_dominance_frontiers cfg doms idoms] returns a map from labels to
+   their frontiers as sets of labels. The definition we use is the following:
+   "the dominance frontier of a node n is the set of all nodes m such that n
+   dominates a predecessor of m, but does not dominate m itself". *)
```
Comment (gretay-js) [inline backend/cfg/cfg_dominators.mli:36]:
```suggestion
(* [compute_dominance_frontiers cfg doms idoms] returns a map from labels to
   their frontiers as sets of labels. The definition we use is the following:
   "the dominance frontier of a node n is the set of all nodes m such that n
   dominates a predecessor of m, but does not strictly dominate m itself". *)
```

**Review by gretay-js — APPROVED** [review]

**Review by mshinwell — APPROVED** [review]
Dune file change only

**Pull request merged**